File: Medium/287. Find the Duplicate Number/287. Find the Duplicate Number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int tortoise = nums[0];
        int hare = nums[0];

        // Phase 1: Find the intersection point of the two pointers
        while (true) {
            tortoise = nums[tortoise];
            hare = nums[nums[hare]];
            if (tortoise == hare) {
                break;
            }
        }

        // Phase 2: Find the entrance to the cycle
        int ptr1 = nums[0];
        int ptr2 = tortoise;

        while (ptr1 != ptr2) {
            ptr1 = nums[ptr1];
            ptr2 = nums[ptr2];
        }

        return ptr1;
    }
}

File: Medium/287. Find the Duplicate Number/287. Find the Duplicate Number.ts
function findDuplicate(nums: number[]): number {
    let slow = nums[0];
    let fast = nums[0];

    // Phase 1: Find the intersection point of the two pointers
    while (true) {
        slow = nums[slow];
        fast = nums[nums[fast]];

        if (slow === fast) {
            break;
        }
    }

    // Phase 2: Find the entrance to the cycle
    let ptr1 = nums[0];
    let ptr2 = slow;

    while (ptr1 !== ptr2) {
        ptr1 = nums[ptr1];
        ptr2 = nums[ptr2];
    }

    return ptr1;
}
